File: software/smart_car_app/lib/service/pke_telemetry.dart
import 'dart:convert';
import 'dart:math';
import 'package:flutter/foundation.dart';

class PkeTelemetrySchema {
  static const String version = 'pke.v1';
}

class PkeTelemetryEvent {
  static const String scanWake = 'scan_wake';
  static const String connect = 'connect';
  static const String auth0Received = 'auth0_received';
  static const String auth1Sent = 'auth1_sent';
  static const String authVerified = 'auth_verified';
  static const String controlFlowAck = 'control_flow_ack';
  static const String rssiSnapshot = 'rssi_snapshot';
  static const String unlockDecision = 'unlock_decision';
}

class PkeTelemetry {
  PkeTelemetry({this.source = 'app'});

  final String source;
  final Random _random = Random();
  int _attemptId = 0;

  int startAttempt({String? vehicleId}) {
    _attemptId = DateTime.now().millisecondsSinceEpoch * 1000 + _random.nextInt(1000);
    emit(
      event: PkeTelemetryEvent.scanWake,
      vehicleId: vehicleId,
      details: 'manual_test_trigger',
    );
    return _attemptId;
  }

  int get attemptId => _attemptId;

  void emit({
    required String event,
    String? vehicleId,
    int? rssiDbm,
    String? unlockDecision,
    String? details,
  }) {
    final payload = <String, dynamic>{
      'schema': PkeTelemetrySchema.version,
      'source': source,
      'attempt_id': _attemptId,
      'event': event,
      'ts_ms': DateTime.now().millisecondsSinceEpoch,
      'vehicle_id': vehicleId,
      'rssi_dbm': rssiDbm,
      'unlock_decision': unlockDecision,
      'details': details,
    };
    debugPrint('[PKE-TLM] ${jsonEncode(payload)}');
  }
}

File: iot/src/ble/pke_telemetry.cpp
#include <Arduino.h>
#include <cstring>

#include "ble/pke_telemetry.h"

namespace {
uint32_t s_attemptId = 0;
char s_vehicleId[9] = {0};

void copyVehicleId(const char* vehicleIdAscii) {
  memset(s_vehicleId, 0, sizeof(s_vehicleId));
  if (!vehicleIdAscii) return;
  strncpy(s_vehicleId, vehicleIdAscii, 8);
  s_vehicleId[8] = '\0';
}
}  // namespace

namespace PKETelemetry {

const char* eventName(Event event) {
  switch (event) {
    case Event::ScanWake: return "scan_wake";
    case Event::Connect: return "connect";
    case Event::Auth0Received: return "auth0_received";
    case Event::Auth1Sent: return "auth1_sent";
    case Event::AuthVerified: return "auth_verified";
    case Event::ControlFlowAck: return "control_flow_ack";
    case Event::RssiSnapshot: return "rssi_snapshot";
    case Event::UnlockDecision: return "unlock_decision";
  }
  return "unknown";
}

void startAttempt(const char* vehicleIdAscii) {
  s_attemptId++;
  copyVehicleId(vehicleIdAscii);
}

uint32_t attemptId() {
  return s_attemptId;
}

void setVehicleId(const char* vehicleIdAscii) {
  copyVehicleId(vehicleIdAscii);
}

void emit(Event event, int rssiDbm, const char* unlockDecision, const char* details) {
  Serial.print("[PKE-TLM] {");
  Serial.print("\"schema\":\"pke.v1\",");
  Serial.print("\"source\":\"esp\",");
  Serial.print("\"attempt_id\":");
  Serial.print(s_attemptId);
  Serial.print(",\"event\":\"");
  Serial.print(eventName(event));
  Serial.print("\",");
  Serial.print("\"ts_ms\":");
  Serial.print((unsigned long)millis());
  Serial.print(",\"vehicle_id\":");
  if (s_vehicleId[0] != '\0') {
    Serial.print("\"");
    Serial.print(s_vehicleId);
    Serial.print("\"");
  } else {
    Serial.print("null");
  }
  Serial.print(",\"rssi_dbm\":");
  if (rssiDbm == kRssiUnknown) {
    Serial.print("null");
  } else {
    Serial.print(rssiDbm);
  }
  Serial.print(",\"unlock_decision\":");
  if (unlockDecision) {
    Serial.print("\"");
    Serial.print(unlockDecision);
    Serial.print("\"");
  } else {
    Serial.print("null");
  }
  Serial.print(",\"details\":");
  if (details) {
    Serial.print("\"");
    Serial.print(details);
    Serial.print("\"");
  } else {
    Serial.print("null");
  }
  Serial.println("}");
}

}  // namespace PKETelemetry
